### PR TITLE
C++: More `cpp/uncontrolled-arithmetic` improvements

### PR DESCRIPTION
**Pull request by MathiasVP**
(Part of https://github.com/github/codeql-c-team/issues/272.)

This PR fixes two sources of false positives I've seen when inspecting the results for `cpp/uncontrolled-arithmetic` on [LGTM](https://lgtm.com/rules/2156240609/alerts/).

- First, we recognize right shifts as a barrier similar to how we recognize division as a barrier. This removes the FPs seen on [love2d/love](https://lgtm.com/projects/g/love2d/love/snapshot/457ba444aa4bfe22d71b396f9f456092e42fa87a/files/src/modules/graphics/ParticleSystem.cpp?sort=name&dir=ASC&mode=heatmap#xe21dd8136e8fdd49:1).
- Second, we check if the `rand()` call is given arguments that explicitly bound the return value. This is rather ad-hoc for now and should be improved by proper modeling of randomness at some point. It does remove many FPs on [gemrb/gemrb](https://lgtm.com/projects/g/gemrb/gemrb/snapshot/1369e9c0464623071fda6d88be2a675d672baae9/files/gemrb/core/Projectile.cpp?sort=name&dir=ASC&mode=heatmap#L157) (querying this project locally we go from 57 results to 0).

### CODE REVIEW

File: cpp/ql/src/Security/CWE/CWE-190/ArithmeticUncontrolled.ql [hunk 1]
```diff
@@ -18,8 +18,16 @@ import semmle.code.cpp.security.TaintTracking
...
+string getAMaxPattern() { result = ["%max%", "%bound%", "h%"] }
+
 predicate isUnboundedRandCall(FunctionCall fc) {
-  fc.getTarget().getName() = "rand" and not bounded(fc)
+  exists(Function func | func = fc.getTarget() |
+    func.getName() = "rand" and
+    not bounded(fc) and
+    not func.getAParameter().getName().toLowerCase().matches([getAMinPattern(), getAMaxPattern()])
+  )
 }
```
Comment (jbj) [inline cpp/ql/src/Security/CWE/CWE-190/ArithmeticUncontrolled.ql:31]: I presume the original code is trying to match only the `int rand(void)` function from `stdlib.h`, not any other function named `rand`. The `rand` function inadvertently matched in `gemrb/gemrb` is a member function of a class. Just switching from `getName` to `hasGlobalOrStdOrBslName` should solve that problem. If you want to further restrict it, you can check that the matched function has no parameters.

Comment (MathiasVP) [reply cpp/ql/src/Security/CWE/CWE-190/ArithmeticUncontrolled.ql:31]: I thought it was actually a feature of the query that we picked any function named `rand` as a possible source of randomness. But looking at the [results](https://lgtm.com/rules/2156240609/alerts/) I see that it's pretty much always a FP whenever it's not `std::rand`. So I'll change the query to use `hasGlobalOrStdOrBslName` instead.

Comment (MathiasVP) [reply cpp/ql/src/Security/CWE/CWE-190/ArithmeticUncontrolled.ql]: Fixed in 615c805b2ce0395b9761e4ca03e472f59c999abe.

**Review by jbj — APPROVED** [review]

**Comment by MathiasVP** [comment]
@jbj I had to add 8765c33847b5b22db8ef0508323bb27a11fb3d0c to keep the new tests happy. Would you mind reapproving?

**Review by jbj — APPROVED** [review]

**Pull request merged**